src/main/java/uk/gov/hmcts/reform/bulkscanprocessor/services/UploadEnvelopeDocumentsService.java

```diff
@@ -0,0 +1,203 @@
+            .collect(groupingBy(Envelope::getContainer))
+            .forEach(this::processByContainer);
+    }
+
+    private void processByContainer(String containerName, List<Envelope> envelopes) {
+        log.info("Processing envelopes in {} container. Envelopes found: {}", containerName, envelopes.size());
+
+        try {
+            CloudBlobContainer blobContainer = blobManager.getContainer(containerName);
+
+            envelopes.forEach(envelope ->
+                getBlobInputStream(
+                    blobContainer,
+                    envelope.getZipFileName()
+                ).flatMap(inputStream ->
+                    processInputStream(inputStream, containerName, envelope.getZipFileName())
+                ).ifPresent(result -> {
+                    boolean isUploaded = uploadParsedZipFileName(envelope, result.getPdfs());
+                    envelopeProcessor.handleEvent(envelope, isUploaded ? DOC_UPLOADED : DOC_UPLOAD_FAILURE);
+                })
+            );
+        } catch (URISyntaxException | StorageException exception) {
+            log.error("Unable to get client for {} container", containerName, exception);
+        }
+    }
+
+    private Optional<BlobInputStream> getBlobInputStream(CloudBlobContainer blobContainer, String zipFileName) {
+        try {
+            return Optional.of(
+                blobContainer
+                    .getBlockBlobReference(zipFileName)
+                    .openInputStream()
+            );
+        } catch (URISyntaxException | StorageException exception) {
+            log.error(
+                "Unable to get blob input stream. Container: {}, blob: {}",
+                blobContainer.getName(),
+                zipFileName
+            );
+
+            return Optional.empty();
+        }
+    }
+
+    private Optional<ZipFileProcessingResult> processInputStream(
+        BlobInputStream blobInputStream,
+        String containerName,
+        String zipFileName
+    ) {
+        try (ZipInputStream zis = new ZipInputStream(blobInputStream)) {
+            return Optional.of(
+                zipFileProcessor.process(zis, containerName, zipFileName)
+            );
+        } catch (DocSignatureFailureException exception) {
+            log.warn(
+                "Rejecting blob - invalid signature. File: {}, Container: {}",
+                zipFileName,
+                containerName,
+                exception
+            );
+
+            handleEventRelatedError(DOC_SIGNATURE_FAILURE, containerName, zipFileName, exception.getMessage());
+
+            blobManager.tryMoveFileToRejectedContainer(zipFileName, containerName, null);
+        } catch (IOException exception) {
+            log.error("Failure reading zip. File: {}, Container: {}", zipFileName, containerName, exception);
+
+            handleEventRelatedError(DOC_FAILURE, containerName, zipFileName, exception.getMessage());
+        }
```

Comment:
I think it would be best to catch Exception instead of IOException here, change the message to a more general one ("Failed to process zip..." or sth) and allow other blobs to be processed.